### PR TITLE
CentCom Galactic Ban DB

### PR DESCRIPTION
**Pull request by ike709**
## About The Pull Request

Ports https://github.com/BeeStation/BeeStation-Hornet/pull/2207

Admins will now be able to look up a player's bans from several other servers via the player panel. My hope is that this will encourage TG admins to make TG's bans public so they can be added.

See: https://centcom.melonmesa.com

Supported servers: 
* BeeStation 
* /vg/station
* FTL13

Planned support that bobbah is working on:
* Yogstation
* Oraclestation
* World Server

## Why It's Good For The Game

It makes the Feweh Global Ban Database a reality.

## Changelog
:cl: ike709 and bobbahbrown
add: Admins can now see your bans on (some) other servers.
/:cl:

### CODE REVIEW

**Comment by ike709** [comment]
image

**Comment by ike709** [comment]
Also it's 4am and I'm about to die of sleep deprivation, please be mindful when reviewing that I may have done dumb things.

**Comment by optimumtact** [comment]
did you try asking the tg admin team first?

**Comment by ike709** [comment]
@optimumtact This doesn't touch TG's ban DB. It just queries the servers that already support the global DB. I don't see why admins *wouldn't* want it. They don't *have* to act on the info that it offers, it's just there to give you more insight on a player.

**Comment by optimumtact** [comment]
Did you try asking the tg admin team first?

**Comment by ike709** [comment]
There I asked them, stop bullying me.

**Comment by lolman360** [comment]
isn't part of the point of opening prs gauging both player, coder and admin feedback

**Comment by optimumtact** [comment]
this PR won't go anywhere unless the admin team and their headadmins approve it, so player or coder feedback hardly matters.

**Comment by optimumtact** [comment]
oh also I guess @MrStonedOne

File: code/modules/admin/topic.dm [hunk 1]
```diff
@@ -2071,6 +2071,53 @@
 
 		usr << browse(dat.Join("<br>"), "window=related_[C];size=420x300")
 
+	else if(href_list["centcomlookup"])
+		if(!check_rights(R_ADMIN))
+			return
+
+		var/ckey = href_list["centcomlookup"]
+
+		var/list/query = json_decode(rustg_http_request_blocking(RUSTG_HTTP_METHOD_GET, "https://centcom.melonmesa.com/ban/search/[ckey]", null, null))
```
Comment (MrStonedOne) [inline code/modules/admin/topic.dm:2080]: This must be configurable, and the server behind this must be open source, and the system must support pushing and pulling from multiple db backends incase everybody just wants to setup their own server and or other trust issues arise.

File: code/modules/admin/topic.dm [hunk 2]
```diff
@@ -2071,6 +2071,53 @@
...
+					dat += "<b>Type: </b> [bandata["type"]]<br>"
+					dat += "<b>Banned By: </b> [bandata["bannedBy"]]<br>"
+					dat += "<b>Reason: </b> [bandata["reason"]]<br>"
+					dat += "<b>Datetime: </b> [bandata["bannedOn"]]<br>"
+					var/expiration = bandata["expires"]
+					dat += "<b>Expires: </b> [expiration ? "[expiration]" : "Permanent"]<br>"
+					if(bandata["type"] == "job")
+						dat += "<b>Jobs: </b> "
+						for(var/job in bandata["jobs"])
+							dat += "[job], "
```
Comment (MrStonedOne) [inline code/modules/admin/topic.dm:2112]: XSS exploit.

A disgruntled admin could inject javascript into this that when viewed run byond commands including admin verbs.

File: code/modules/admin/topic.dm [hunk 3]
```diff
@@ -2071,6 +2071,53 @@
...
+						dat += "<br>"
+					dat += "<hr>"
+
+		dat += "<br></body>"
+		var/datum/browser/popup = new(usr, "centcomlookup-[ckey]", "<div align='center'>Central Command Galactic Ban Database</div>", 700, 600)
+		popup.set_content(dat)
+		popup.open(0)
```
Comment (MrStonedOne) [inline code/modules/admin/topic.dm:2119]: eww browser datum.

(ok, i get why you did it, but eww =P)

File: code/modules/admin/topic.dm (hunk 1 above)
Comment (ike709) [reply code/modules/admin/topic.dm:2080]: image

**Comment by ike709** [comment]
Wrapped everything in `sanitize()`, made the URL a config, and disabled it by default.

Bobbah is working on MSO's other requirements.

**Comment by bobbahbrown** [comment]
This is true that I am working on these things, it'll take me a short while because I'm super slammed with stuff this week. Hoping to get a chance to put that together this week.

File: code/modules/admin/topic.dm [hunk 4]
```diff
@@ -2071,6 +2071,57 @@
...
+		if(!check_rights(R_ADMIN))
+			return
+
+		if(!CONFIG_GET(string/centcom_ban_db))
+			to_chat(usr, "<span class='warning'>Centcom Galactic Ban DB is disabled!</span>")
+			return
+
+		var/ckey = href_list["centcomlookup"]
+
+		var/list/query = json_decode(rustg_http_request_blocking(RUSTG_HTTP_METHOD_GET, "[CONFIG_GET(string/centcom_ban_db)][ckey]", null, null))
```
Comment (Cyberboss) [inline code/modules/admin/topic.dm:2084]: Use the [http request helper datum](https://github.com/tgstation/tgstation/blob/master/code/datums/http.dm) and make it async please. Frankly, I never want to see this version of the function used. We don't need http calls locking up DD.

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/modules/admin/topic.dm [hunk 5]
```diff
@@ -2083,7 +2083,7 @@
 
 		// Make the request
 		var/datum/http_request/request = new()
-		request.prepare(RUSTG_HTTP_METHOD_GET, "[CONFIG_GET(string/centcom_ban_db)][ckey]", "", "")
+		request.prepare(RUSTG_HTTP_METHOD_GET, "[CONFIG_GET(string/centcom_ban_db)]/[ckey]", "", "")
 		request.begin_async()
 		UNTIL(request.is_complete())
```
Comment (Cyberboss) [inline code/modules/admin/topic.dm:2088]:
```suggestion
		UNTIL(request.is_complete() || !usr)
		if (!usr)
			return
```

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by ike709** [comment]
BYOND is better at parsing JSON than I thought it was, no more parsing shenanigans with `splittext()`.

File: code/modules/admin/topic.dm [hunk 6]
```diff
@@ -2071,6 +2071,59 @@
...
+		request.prepare(RUSTG_HTTP_METHOD_GET, "[CONFIG_GET(string/centcom_ban_db)]/[ckey]", "", "")
+		request.begin_async()
+		UNTIL(request.is_complete() || !usr)
+		if (!usr)
+			return
+		var/datum/http_response/response = request.into_response()
+
+		var/list/bans
+
+		var/dat = "<meta http-equiv='Content-Type' content='text/html; charset=UTF-8'><body>"
```
Comment (Rohesie) [inline code/modules/admin/topic.dm:2095]:
```suggestion
		var/list/dat = list("<meta http-equiv='Content-Type' content='text/html; charset=UTF-8'><body>")
```

File: code/modules/admin/topic.dm [hunk 7]
```diff
@@ -2071,6 +2071,59 @@
...
+					if(ban["type"] == "job")
+						dat += "<b>Jobs: </b> "
+						var/list/jobs = ban["jobs"]
+						dat += sanitize(jobs.Join(", "))
+						dat += "<br>"
+					dat += "<hr>"
+
+		dat += "<br></body>"
+		var/datum/browser/popup = new(usr, "centcomlookup-[ckey]", "<div align='center'>Central Command Galactic Ban Database</div>", 700, 600)
+		popup.set_content(dat)
```
Comment (Rohesie) [inline code/modules/admin/topic.dm:2124]:
```suggestion
		popup.set_content(dat.Join())
```

**Review by Cyberboss — APPROVED** [review]

**Comment by bobbahbrown** [comment]
@MrStonedOne https://github.com/bobbahbrown/CentCom

tada!

**Pull request merged**